source/common/http/conn_manager_utility.cc

```diff
@@ -127,12 +127,24 @@ Network::Address::InstanceConstSharedPtr ConnectionManagerUtility::mutateRequest
           connection.ssl() ? Headers::get().SchemeValues.Https : Headers::get().SchemeValues.Http);
     }
   } else {
-    // If we are not using remote address, attempt to pull a valid IPv4 or IPv6 address out of XFF.
+    // If we are not using remote address, attempt to pull a valid IPv4 or IPv6 address out of XFF
+    // or through an extension. An extension might be needed when XFF doesn't work (e.g. an
+    // irregular network).
+    //
     // If we find one, it will be used as the downstream address for logging. It may or may not be
     // used for determining internal/external status (see below).
-    auto ret = Utility::getLastAddressFromXFF(request_headers, xff_num_trusted_hops);
-    final_remote_address = ret.address_;
-    single_xff_address = ret.single_address_;
+    auto original_ip_detection = config.originalIpDetection();
+    if (original_ip_detection) {
+      struct OriginalIPDetectionParams params = {request_headers};
+      final_remote_address = original_ip_detection->detect(params);
+    }
+
+    // If there's no extension or it failed to detect, give XFF a try.
```

Comment:
we should definitely comment this up in the include docs, and make it clear that the extension is called ahead of XFF checks, but fails over to standard mechanisms.

Reply:
Added some to `docs/root/intro/arch_overview/other_features/ip_transparency.rst`, the header file and the proto files.

Reply:
Alternatively, should we consider moving XFF handling to an extension that implements the API that is added in this PR.

Reply:
I was going to suggest the same thing that @antoniovicente did.  If we're adding an extension point here, let's take the opportunity to move the existing functionality into an extension, and just make that the default extension.